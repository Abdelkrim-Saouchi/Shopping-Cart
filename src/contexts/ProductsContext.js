import { createContext } from 'react';

export const ProductsContext = createContext({
  addProductToCart: null,
  deleteProductFromCart: null,
});
