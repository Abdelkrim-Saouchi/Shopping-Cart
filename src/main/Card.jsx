const Card = ({ imgSrc, productName, price }) => {
  return (
    <div className="card">
      <img src={imgSrc} alt="product img" />
      <div className="text">
        <h3>{productName}</h3>
        <p>{price}$</p>
      </div>
      <button>Add to Cart</button>
    </div>
  );
};

export default Card;
