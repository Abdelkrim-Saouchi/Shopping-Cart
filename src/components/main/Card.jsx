import { useContext, useState } from 'react';
import { ProductsContext } from '../../contexts/ProductsContext';
import styles from './Card.module.css';

const Card = ({ id, imgSrc, productName, price }) => {
  const { addProductToCart, deleteProductFromCart } =
    useContext(ProductsContext);
  const [quantity, setQuantity] = useState(0);
  const [added, setAdded] = useState(false);

  const incrementQuantity = () => {
    setQuantity(quantity + 1);
  };

  const decrementQuantity = () => {
    if (quantity > 0) {
      setQuantity(quantity - 1);
    }
  };

  const changeQuantity = (e) => {
    const quantity = Number(e.target.value);
    if (!isNaN(quantity)) setQuantity(quantity);
  };

  return (
    <div className={styles.card}>
      <img src={imgSrc} alt="product" />
      <div className={styles.text}>
        <h3>{productName}</h3>
        <p>{price}$</p>
      </div>
      <div className={styles.quantity}>
        <button
          className="decrement"
          aria-label="decrement quantity"
          onClick={decrementQuantity}
        >
          -
        </button>
        <span id={'label' + id}>Quantity:</span>
        <input
          type="text"
          id={'product_quantity' + id}
          aria-labelledby={'label' + id}
          data-testid="product_quantity"
          value={quantity}
          onChange={changeQuantity}
        />
        <button
          className="increment"
          aria-label="increment quantity"
          onClick={incrementQuantity}
        >
          +
        </button>
      </div>
      {quantity > 0 ? (
        <button
          className={styles.addToCart}
          onClick={() => {
            if (added) {
              setAdded(false);
              deleteProductFromCart(id);
            } else {
              setAdded(true);
              addProductToCart(id, productName, imgSrc, price, quantity);
            }
          }}
        >
          {added ? 'Added to Cart' : 'Add to Cart'}
        </button>
      ) : (
        <button
          className={styles.addToCart}
          onClick={() => {
            if (added) {
              setAdded(false);
              deleteProductFromCart(id);
            }
          }}
        >
          {added ? 'Added to Cart' : 'Add to Cart'}
        </button>
      )}
    </div>
  );
};

export default Card;
