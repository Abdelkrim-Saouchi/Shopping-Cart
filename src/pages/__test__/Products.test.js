import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import Card from '../../components/main/Card';
import { ProductsContext } from '../../contexts/ProductsContext';

const renderWithRouter = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    ...render(ui, { wrapper: BrowserRouter }),
  };
};

let providerProps = {
  value: { addProductToCart: jest.fn(), deleteProductFromCart: jest.fn() },
};

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(
    <ProductsContext.Provider {...providerProps}>
      {ui}
    </ProductsContext.Provider>,
    renderOptions
  );
};

describe('Card Component', () => {
  const mockProps = {
    id: 1,
    imgSrc: 'img.jpg',
    productName: 'product',
    price: 10,
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('Card renders product info correctly', () => {
    const { container } = renderWithRouter(<Card {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  test('increments quantity when increment button is clicked', () => {
    renderWithRouter(<Card {...mockProps} />);

    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);

    expect(screen.getByTestId('product_quantity').value).toBe('1');
  });

  test('decrements quantity when decrement button is clicked', () => {
    renderWithRouter(<Card {...mockProps} />);

    const incrementButton = screen.getByText('+');
    const decrementButton = screen.getByText('-');
    fireEvent.click(incrementButton);
    fireEvent.click(decrementButton);

    expect(screen.getByTestId('product_quantity').value).toBe('0');
  });

  test('calls addProductToCart when Add to Cart button is clicked with quantity > 0', () => {
    customRender(<Card {...mockProps} />, { providerProps });
    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);
    const addToCartButton = screen.getByText('Add to Cart');
    fireEvent.click(addToCartButton);

    expect(providerProps.value.addProductToCart).toHaveBeenCalledWith(
      1,
      'product',
      'img.jpg',
      10,
      1
    );
  });

  test('calls deleteProductFromCart when Add to Cart button is clicked with quantity > 0 and already added', () => {
    customRender(<Card {...mockProps} />, { providerProps });
    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);

    const addToCartButton = screen.getByText('Add to Cart');
    fireEvent.click(addToCartButton);

    const addedToCartButton = screen.getByText('Added to Cart');
    fireEvent.click(addedToCartButton);

    expect(providerProps.value.deleteProductFromCart).toHaveBeenCalledWith(1);
  });

  test('calls deleteProductFromCart when Add to Cart button is clicked with quantity = 0 and already added', () => {
    customRender(<Card {...mockProps} />, { providerProps });

    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);

    const addToCartButton = screen.getByText('Add to Cart');
    fireEvent.click(addToCartButton);

    const decrementButton = screen.getByText('-');
    fireEvent.click(decrementButton);

    const addedToCartButton = screen.getByText('Added to Cart');
    fireEvent.click(addedToCartButton);

    expect(providerProps.value.deleteProductFromCart).toHaveBeenCalledWith(1);
  });

  test('does not call addProductToCart or deleteProductFromCart when Add to Cart button is clicked with quantity = 0 and not added', () => {
    customRender(<Card {...mockProps} />, { providerProps });
    const addToCartButton = screen.getByText('Add to Cart');
    const quantityInput = screen.getByTestId('product_quantity');

    fireEvent.change(quantityInput, { target: { value: '0' } });
    fireEvent.click(addToCartButton);

    expect(providerProps.value.addProductToCart).not.toHaveBeenCalled();
    expect(providerProps.value.deleteProductFromCart).not.toHaveBeenCalled();
  });
});
