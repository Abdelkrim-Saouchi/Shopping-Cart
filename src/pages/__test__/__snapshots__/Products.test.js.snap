// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component Card renders product info correctly 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="product"
      src="img.jpg"
    />
    <div
      class="text"
    >
      <h3>
        product
      </h3>
      <p>
        10
        $
      </p>
    </div>
    <div
      class="quantity"
    >
      <button
        aria-label="decrement quantity"
        class="decrement"
      >
        -
      </button>
      <span
        id="label1"
      >
        Quantity:
      </span>
      <input
        aria-labelledby="label1"
        data-testid="product_quantity"
        id="product_quantity1"
        type="text"
        value="0"
      />
      <button
        aria-label="increment quantity"
        class="increment"
      >
        +
      </button>
    </div>
    <button
      class="addToCart"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
